teps/0058-graceful-pipeline-run-termination.md

```diff
@@ -0,0 +1,198 @@
+---
+status: proposed
+title: Graceful Pipeline Run Termination
+creation-date: '2021-03-18'
+last-updated: '2021-03-18'
+authors:
+- '@@rafalbigaj'
+---
+
+# TEP-0058: Graceful Pipeline Run Termination
+---
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+  - [Use Cases (optional)](#use-cases-optional)
+- [Requirements](#requirements)
+- [Proposal](#proposal)
+  - [Notes/Caveats (optional)](#notescaveats-optional)
+  - [Risks and Mitigations](#risks-and-mitigations)
+  - [User Experience (optional)](#user-experience-optional)
+  - [Performance (optional)](#performance-optional)
+- [Design Details](#design-details)
+- [Test Plan](#test-plan)
+- [Design Evaluation](#design-evaluation)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+- [References (optional)](#references-optional)
+<!-- /toc -->
+
+## Summary
+
+Marking a `PipelineRun` as [cancelled](https://github.com/tektoncd/pipeline/blob/main/docs/pipelineruns.md#cancelling-a-pipelinerun)
+stops all running tasks and deletes associated Pods. That prevents final tasks, specified in a `Pipeline` under `finally` section,
+from being run.
+There should be a way to terminate a `PipelineRun` gracefully and wait for cleanup actions triggered by final tasks.
+
+## Motivation
+
+It is common that tasks when run trigger execution of external activities or request resources in foreign systems.
+Final tasks are perfectly suited for any cleanup operations that have to be performed when execution is completed.
+That is required in case of success, failure, as well as pipeline run cancellation.
+ 
+Currently, there is no way to terminate a `PipelineRun` gracefully.
+The existing cancellation capability is problematic in the real use cases.
```

Comment:
💯 we have seen the same problem here. The expectation with a `finally` is that it runs whether a pipeline is cancelled, successful or whatever.

Reply:
I agree, a graceful finally would be really helpful!

We must also preserve the existing functionality to avoid a breaking change, so ideally we will end up with "forced" cancel, which is what we have today, and a graceful one, which is what is proposed here today.

Reply:
Yup, leave the pipeline run cancel as is, cancelling the entire `pipelineRun`, need a state to terminate the `tasks` section.